pallets/teebag/src/lib.rs

```diff
@@ -0,0 +1,833 @@
+		MaxEnclaveCountUnderflow,
+		/// A proposed block is unexpected.
+		ReceivedUnexpectedSidechainBlock,
+		/// The value for the next finalization candidate is invalid.
+		InvalidNextFinalizationCandidateBlockNumber,
+	}
+
+	#[pallet::storage]
+	#[pallet::getter(fn admin)]
+	pub type Admin<T: Config> = StorageValue<_, T::AccountId, OptionQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn mode)]
+	pub type Mode<T: Config> = StorageValue<_, OperationalMode, ValueQuery>;
+
+	#[pallet::type_value]
+	pub fn DefaultMaxEnclaveCount() -> u64 {
+		3
+	}
+
+	#[pallet::storage]
+	#[pallet::getter(fn max_enclave_count)]
+	pub type MaxEnclaveCount<T: Config> =
+		StorageMap<_, Blake2_128Concat, WorkerType, u64, ValueQuery, DefaultMaxEnclaveCount>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn enclave_count)]
+	pub type EnclaveCount<T: Config> = StorageMap<_, Blake2_128Concat, WorkerType, u64, ValueQuery>;
+
+	// registry that holds all registered enclaves, using T::AccountId as the key
+	// having `worker_type` and `mrenclave` in each `Enclave` instance might seem a bit redundant,
+	// but it increases flexibility where we **could** allow the same type of worker to have
+	// different mrenclaves - e.g. when more than one version of an enclave is permitted in TEE-node
+	// cluster.
+	//
+	// It simplifies the lookup a bit too, otherwise we might need several storages.
+	#[pallet::storage]
+	#[pallet::getter(fn enclave_registry)]
+	pub type EnclaveRegistry<T: Config> =
+		StorageMap<_, Blake2_128Concat, T::AccountId, Enclave, OptionQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn allow_sgx_debug_mode)]
+	pub type AllowSGXDebugMode<T: Config> = StorageValue<_, bool, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn quoting_enclave_registry)]
+	pub type QuotingEnclaveRegistry<T: Config> = StorageValue<_, QuotingEnclave, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn tcb_info)]
+	pub type TcbInfo<T: Config> =
+		StorageMap<_, Blake2_128Concat, Fmspc, TcbInfoOnChain, ValueQuery>;
+
+	// keep track of a list of scheduled/allowed enchalves, mainly used for enclave updates,
+	// can only be modified by AdminOrigin
+	// (worker_type, sidechain_block_number) -> expected MrEnclave
+	//
+	// about the first time enclave registration:
+	// prior to `register_enclave` this map needs to be populated with ((worker_type, 0),
+	// expected-mrenclave), otherwise the registration will fail
+	//
+	// For NON-sidechain worker_type, we still use this storage to whitelist mrenclave, in this case
+	// the `SidechainBlockNumber` is ignored - you could always set it to 0.
+	//
+	// Theorectically we could always push the enclave in `register_enclave`, but we want to
+	// limit the mrenclave that can be registered, as the parachain is supposed to process enclaves
+	// with specific mrenclaves.
+	#[pallet::storage]
+	#[pallet::getter(fn scheduled_enclave)]
+	pub type ScheduledEnclave<T: Config> =
```

Comment:
I think it doesn't scale well.
We cannot have bitacross and identity worker both running in sidechain mode, can we ?
Each worker should have an identifier (`identity`, `bitacross`) and (WorkerIdentifier, SidechainBlockNumber) should be used as a key here. what do you think?

Reply:
With "scale well" do you mean we should have more storages to facilitate the lookup?

> Each worker should have an identifier (identity, bitacross)

It's there as a field in `Enclave` struct

> (WorkerIdentifier, SidechainBlockNumber)

"WorkerIdentifier" is `T::AccountId` here, what exactly does `SidechainBlockNumber` do?

> We cannot have bitacross and identity worker both running in sidechain mode, can we ?

Not now, but maybe later (just in case they want to manage some common state).
Either way we need features of a) mrenclave whitelists b) for sidechain workers, schedule mrenclave for upgrade

Reply:
By scalability I mean we cannot run both bitacross and identity worker in sidechain mode in such setup. Just pointing this out :) but maybe that's fine for the first version.


> "WorkerIdentifier" is T::AccountId here, what exactly does SidechainBlockNumber do?

`SidechainBlockNumber` indicates sidechain block from which new mrenclave is required. I just replaced `WorkerType` with `WorkerIdentifier` in the tuple. `WorkerIdentifier` could be `identity` | `bitacross`